packages/bsky/tests/views/notifications.test.ts

```diff
@@ -419,6 +419,73 @@ describe('notification views', () => {
     expect(forSnapshot(notifs.data)).toMatchSnapshot()
   })
 
+  it('filters notifications by reason', async () => {
+    const res = await agent.app.bsky.notification.listNotifications(
+      {
+        filter: ['mention'],
+      },
+      {
+        headers: await network.serviceHeaders(
+          sc.dids.alice,
+          ids.AppBskyNotificationListNotifications,
+        ),
+      },
+    )
+    expect(res.data.notifications.length).toBe(1)
+    expect(forSnapshot(res.data)).toMatchSnapshot()
+  })
+
+  it('filters notifications by multiple reasons', async () => {
+    const res = await agent.app.bsky.notification.listNotifications(
+      {
+        filter: ['mention', 'reply'],
+      },
+      {
+        headers: await network.serviceHeaders(
+          sc.dids.alice,
+          ids.AppBskyNotificationListNotifications,
+        ),
+      },
+    )
+    expect(res.data.notifications.length).toBe(4)
+    expect(forSnapshot(res.data)).toMatchSnapshot()
+  })
+
+  it('paginates filtered notifications', async () => {
+    const results = (results) =>
+      sort(results.flatMap((res) => res.notifications))
+    const paginator = async (cursor?: string) => {
+      const res = await agent.app.bsky.notification.listNotifications(
+        { filter: ['mention', 'reply'], cursor, limit: 2 },
+        {
+          headers: await network.serviceHeaders(
+            alice,
+            ids.AppBskyNotificationListNotifications,
+          ),
+        },
+      )
+      return res.data
+    }
+
+    const paginatedAll = await paginateAll(paginator)
+    paginatedAll.forEach((res) =>
+      expect(res.notifications.length).toBeLessThanOrEqual(2),
```

Comment:
Actually <=2 is correct. The semantic that we use for `limit` is "up to this number but it can be less". In cases of results getting stripped out for blocks/mutes etc it may be less than the limit. In this particular case, we paginate on the filtered result set until we get 1/2 of the limit (so in this case 1 result)